### PR TITLE
Feature/swap UI

### PR DESCRIPTION
**Pull request by evbambly**
Enabling Swapping a current donation with another one.

Logic is listed [here](https://docs.google.com/document/d/1oyXCdA9WQScMRx1DwYgc8q2LiFfcJzgKPqLQUCK-RRE/edit?usp=sharing)

UI is depicted [here](https://www.figma.com/file/LqRIEuQgMG8Kd4IZRv40rd/zm-racaz?node-id=1513%3A1512)

Trello task is [here](https://trello.com/c/5B8WhTN5/304-donor-swap-appointments-ui)

### CODE REVIEW

File: packages/donor/src/components/AppHeader/AppHeader.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ export default function AppHeader({
   if (hasBurgerMenu) {
     icon = (
       <IconButton
-        buttonName="burger_menu"
+        analyticsName="burger_menu"
```
Comment (yaron1m) [inline packages/donor/src/components/AppHeader/AppHeader.tsx:52]: Why is this renaming part of this PR?

Please split this PR to a refactor one and a UI one

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx [hunk 2]
```diff
@@ -15,9 +15,11 @@ const props: QuestionnaireScreenProps = {
   donationStartTimeMillis: 1628845200000,
   onBack: action("onBack"),
   onSuccess: action("onSuccess"),
-  goToHomePage: async () => action("goToHomePage")(),
+  goBackAndRefresh: async () => action("goToHomePage")(),
```
Comment (yaron1m) [inline packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx:18]: Same for this

File: packages/donor/src/components/AppointmentPicker/AppointmentPicker.tsx [hunk 3]
```diff
@@ -16,12 +16,14 @@ export interface AppointmentPickerProps {
   donationDay: DonationDay;
   onSlotSelected: (donationSlot: DonationSlotToBook) => void;
   showHospitalName: boolean;
+  isSwapAppointment: boolean;
```
Comment (yaron1m) [inline packages/donor/src/components/AppointmentPicker/AppointmentPicker.tsx:19]: Name is not clear.
What does this field mean? What is it used for? What effect does it have?

**Review by yaron1m — COMMENTED** [review]
Please make sure all actions run before I review this PR, so that the Storybook would have the latest version deployed.

File: packages/donor/src/components/AppHeader/AppHeader.tsx (hunk 1 above)
Comment (evbambly) [reply packages/donor/src/components/AppHeader/AppHeader.tsx]: Done

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx (hunk 2 above)
Comment (evbambly) [reply packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx]: Done

File: packages/donor/src/components/AppointmentPicker/AppointmentPicker.tsx (hunk 3 above)
Comment (evbambly) [reply packages/donor/src/components/AppointmentPicker/AppointmentPicker.tsx]: Clarified and cut unnecessary instances of this variable name.

File: packages/donor/src/screens/bookDonation/BookDonationScreen.tsx [hunk 4]
```diff
@@ -67,116 +40,13 @@ export default function BookDonationScreen({
...
-
-      <div className={styles.donationsCard}>
-        {Donations(
-          selectedHospital,
-          isFetching,
-          sortedDonationDays,
-          onSlotSelected
-        )}
-      </div>
+      <AppointmentSelect
```
Comment (yaron1m) [inline packages/donor/src/screens/bookDonation/BookDonationScreen.tsx:43]: Same for this.
Can the extraction of the appointment selection to a component be in a different PR?

The reason is that reviewing a PR with 50 files and 4,601 lines changes is very hard. 

Please go over this entire pr and try to split it to smaller commits. 
Anything that is not pure Appointment Swap feature should be in a different PR.

File: packages/donor/src/screens/questionnaire/QuestionnaireScreenContainer.tsx [hunk 5]
```diff
@@ -52,42 +48,16 @@ export function QuestionnaireScreenContainer(
...
+        props.setBookedAppointment(response.bookedAppointment)
       );
-
-    switch (bookAppointmentResponse.status) {
-      case FunctionsApi.BookAppointmentStatus.NO_AVAILABLE_APPOINTMENTS:
-        setError(bookAppointmentResponse.status);
-        setIsLoading(false);
-        break;
-
-      case FunctionsApi.BookAppointmentStatus.SUCCESS:
```
Comment (yaron1m) [inline packages/donor/src/screens/questionnaire/QuestionnaireScreenContainer.tsx:74]: Same here, for example.
Extracting this to a dedicated hook is a nice idea, but should be in a dedicated PR.

File: packages/donor/src/screens/bookDonation/BookDonationScreen.tsx (hunk 4 above)
Comment (evbambly) [reply packages/donor/src/screens/bookDonation/BookDonationScreen.tsx:43]: Rolled Back

File: packages/donor/src/screens/questionnaire/QuestionnaireScreenContainer.tsx (hunk 5 above)
Comment (evbambly) [reply packages/donor/src/screens/questionnaire/QuestionnaireScreenContainer.tsx]: Ran a sweep on the entire PR, 
I don't believe there is any refactoring left in this PR.

File: packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    return (
+      <div className={styles.loading}>
+        <Player
+          autoplay
+          loop
+          src={DropAnimation}
+          className={styles.dropAnimation}
+        />
+        <div className={styles.loadingText}>מחפש תורים קרובים...</div>
+      </div>
```
Comment (yaron1m) [inline packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:126]: This should be a standalone component (used here and in other places).

File: packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  appointmentToHide,
+  isSwapAppointment,
+}: AppointmentSelectProps) {
+  const [selectedHospital, setSelectedHospital] = useState<Hospital | "">("");
+
+  const sortedDonationDays = React.useMemo(() => {
+    const appointmentFilter = (appointment: AvailableAppointment) => {
+      const isValidHospital =
+        appointment.hospital === selectedHospital || !selectedHospital;
+
```
Comment (yaron1m) [inline packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:49]: if isValidHospital == false, there's no point to continue, just return false.

File: packages/donor/src/components/DonationInfo/DonationInfo.tsx [hunk 8]
```diff
@@ -1,24 +1,27 @@
...
 import { DateUtils, Hospital, LocaleUtils } from "@zm-blood-components/common";
-import React from "react";
 
 export interface DonationToBookInfoProps {
   donationStartTimeMillis: number;
   hospital: Hospital;
+  isPreviousAppointmentInfo: boolean;
 }
 
 export default function DonationToBookInfo({
```
Comment (yaron1m) [inline packages/donor/src/components/DonationInfo/DonationInfo.tsx:11]: If you rename the file, please rename the function and the props.

File: packages/donor/src/components/DonationInfo/DonationInfo.stories.tsx [hunk 9]
```diff
@@ -1,23 +1,22 @@
...
+  component: DonationInfo,
   title: "Components/Donation to Book",
   parameters: { layout: "fullscreen" },
 };
 
-const props: DonationToBookInfoProps = {
+const props: DonationInfoProps = {
   donationStartTimeMillis: 1625823300000,
   hospital: Hospital.ICHILOV,
+  isPreviousAppointmentInfo: false,
```
Comment (yaron1m) [inline packages/donor/src/components/DonationInfo/DonationInfo.stories.tsx:16]: I am missing another story where this field is true.

File: packages/donor/src/components/DonationInfo/DonationInfo.tsx [hunk 10]
```diff
@@ -1,24 +1,27 @@
-import styles from "./DonationToBookInfo.module.scss";
+import styles from "./DonationInfo.module.scss";
 import Calendar from "../../assets/images/AppointmentCalendar.svg";
 import { DateUtils, Hospital, LocaleUtils } from "@zm-blood-components/common";
-import React from "react";
 
 export interface DonationToBookInfoProps {
   donationStartTimeMillis: number;
   hospital: Hospital;
+  isPreviousAppointmentInfo: boolean;
```
Comment (yaron1m) [inline packages/donor/src/components/DonationInfo/DonationInfo.tsx:8]: 1. I think you mean existing, not previous.
2. This is a component, and a field like `isPreviousAppointmentInfo` is very coupled with the domain logic. Please try a name that does not leak this logic inside. Either you rename the field, or take the entire title as a string parameter.

File: packages/donor/src/components/basic/IconButton/IconButton.tsx [hunk 11]
```diff
@@ -3,23 +3,27 @@ import {
...
-  onClick,
-  ...props
-}: IconButtonProps) {
-  const handleClick = () => {
-    onClick();
+export default function IconButton({ buttonName, ...props }: IconButtonProps) {
+  const handleClick = (
+    e: MouseEvent<HTMLButtonElement, globalThis.MouseEvent>
+  ) => {
+    props.onClick?.(e);
```
Comment (yaron1m) [inline packages/donor/src/components/basic/IconButton/IconButton.tsx:18]: Please explain this change

File: packages/donor/src/components/basic/Menu/Menu.tsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
+import {
+  Menu as MuiMenu,
+  MenuProps as MuiMenuProps,
+  MenuItem as MuiMenuItem,
+} from "@mui/material";
+import { AnalyticsButtonType } from "@zm-blood-components/common";
+import { ReactElement } from "react";
+import { reportClick } from "../../../Analytics";
+
+export interface MenuProps extends MuiMenuProps {
```
Comment (yaron1m) [inline packages/donor/src/components/basic/Menu/Menu.tsx:10]: I'm not a fan of extending an external props object.
I prefer listing all the fields I need here, one by one.
This makes writing the stories much easier, and helps other developers understand what capabilities are used in this project.

File: packages/donor/src/components/basic/Menu/Menu.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+export const Default: Story<MenuProps> = (args) => {
+  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);
+
+  return (
+    <div>
+      <IconButton
+        buttonName="open_menu"
+        onClick={(e) => setAnchorEl(e.currentTarget)}
```
Comment (yaron1m) [inline packages/donor/src/components/basic/Menu/Menu.stories.tsx:20]: Why do you need `Typography` here?

File: packages/donor/src/components/popups/BookingAppointmentErrorPopup.tsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import Popup from "../basic/Popup";
+import { FunctionsApi } from "@zm-blood-components/common";
+
+interface BookingAppointmentErrorPopupProps {
+  errorCode?: FunctionsApi.BookAppointmentStatus;
+  onApproved: () => Promise<void>;
+}
+
+export default function BookingAppointmentErrorPopup({
```
Comment (yaron1m) [inline packages/donor/src/components/popups/BookingAppointmentErrorPopup.tsx:9]: Where are the stories?

File: packages/donor/src/components/popups/ConfirmSwapAppointmentPopup.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+import SwapLogo from "../../assets/images/swap_popup.svg";
+
+interface ConfirmSwapAppointmentPopupProps {
+  isOpen: boolean;
+  onApproved: () => void | Promise<void>;
+  onBack: () => void;
+  onClose: () => void;
+}
+
+export default function ConfirmSwapAppointmentPopup({
```
Comment (yaron1m) [inline packages/donor/src/components/popups/ConfirmSwapAppointmentPopup.tsx:11]: Stories please

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx [hunk 16]
```diff
@@ -18,6 +18,8 @@ const props: QuestionnaireScreenProps = {
   goToHomePage: async () => action("goToHomePage")(),
   isLoading: false,
   debugMode: false,
+  isSwapAppointment: false,
+  bookingErrorCode: undefined,
```
Comment (yaron1m) [inline packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx:22]: No need to add an undefined field

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.tsx [hunk 17]
```diff
@@ -1,50 +1,62 @@
...
+import BookingAppointmentErrorPopup from "../../components/popups/BookingAppointmentErrorPopup";
 
 export interface QuestionnaireScreenProps {
   hospital: Hospital;
   donationStartTimeMillis: number;
   onSuccess: () => void;
   isLoading: boolean;
   debugMode: boolean;
-  errorCode?: FunctionsApi.BookAppointmentStatus;
+  bookingErrorCode: FunctionsApi.BookAppointmentStatus | undefined;
```
Comment (yaron1m) [inline packages/donor/src/screens/questionnaire/QuestionnaireScreen.tsx:19]: Why not use the `?:` syntax?

**Review by yaron1m — COMMENTED** [review]
I didn't read all the files, but some comments are ready, so I decided to submit them now.

More to follow :)

File: packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+export interface AppointmentSelectProps {
+  availableAppointments: AvailableAppointment[];
+  isFetching: boolean;
+  onSlotSelected: (donationSlot: DonationSlotToBook) => void;
+  tooCloseDonationPopupProps: Pick<
+    PopupProps,
+    "open" | "onBack" | "onApproved" | "onClose"
+  >;
+  appointmentToHide?: BookedAppointment;
```
[the remark this answers is not in the document]
Comment (evbambly) [reply packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:31]: It is currently booked,
if the process is terminated, 
it will remain booked.

[the remark this answers is not in the document]
Comment (evbambly) [reply packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:31]: Done

File: packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx (hunk 6 above)
Comment (evbambly) [reply packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:126]: This is something I moved over to the refactor PR.
Should I return it?

File: packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx (hunk 7 above)
Comment (evbambly) [reply packages/donor/src/components/AppointmentSelect/AppointmentSelect.tsx:49]: Done

File: packages/donor/src/components/DonationInfo/DonationInfo.tsx (hunk 8 above)
Comment (evbambly) [reply packages/donor/src/components/DonationInfo/DonationInfo.tsx:11]: Done

File: packages/donor/src/components/DonationInfo/DonationInfo.stories.tsx (hunk 9 above)
Comment (evbambly) [reply packages/donor/src/components/DonationInfo/DonationInfo.stories.tsx:16]: Done

File: packages/donor/src/components/DonationInfo/DonationInfo.tsx (hunk 10 above)
Comment (evbambly) [reply packages/donor/src/components/DonationInfo/DonationInfo.tsx:8]: Done

File: packages/donor/src/components/basic/IconButton/IconButton.tsx (hunk 11 above)
Comment (evbambly) [reply packages/donor/src/components/basic/IconButton/IconButton.tsx:18]: This was unnecessary.
I undid the change and converted UpcomingDonationScreen Line #44 to work with React Ref.

File: packages/donor/src/components/basic/Menu/Menu.tsx (hunk 12 above)
Comment (evbambly) [reply packages/donor/src/components/basic/Menu/Menu.tsx:10]: In this case, I agree.
Done

File: packages/donor/src/components/basic/Menu/Menu.stories.tsx (hunk 13 above)
Comment (evbambly) [reply packages/donor/src/components/basic/Menu/Menu.stories.tsx:20]: Removed

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx (hunk 16 above)
Comment (evbambly) [reply packages/donor/src/screens/questionnaire/QuestionnaireScreen.stories.tsx:22]: Done

File: packages/donor/src/screens/questionnaire/QuestionnaireScreen.tsx (hunk 17 above)
Comment (evbambly) [reply packages/donor/src/screens/questionnaire/QuestionnaireScreen.tsx:19]: Done

File: packages/donor/src/components/popups/BookingAppointmentErrorPopup.tsx (hunk 14 above)
Comment (evbambly) [reply packages/donor/src/components/popups/BookingAppointmentErrorPopup.tsx:9]: Done

File: packages/donor/src/components/popups/ConfirmSwapAppointmentPopup.tsx (hunk 15 above)
Comment (evbambly) [reply packages/donor/src/components/popups/ConfirmSwapAppointmentPopup.tsx:11]: Done

**Comment by evbambly** [comment]
As per our conversation, @yaron1m ,
this code works and I'll merge it now :)
